src/main/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImpl.java

```diff
@@ -302,6 +328,70 @@ public List<StudyEntryColumnDTO> getStudyEntryColumns(final Integer studyId, fin
 		return this.middlewareStudyEntryService.getStudyEntryColumns(studyId, programUUID);
 	}
 
+	@Override
+	public List<Variable> getVariableListByStudyAndType(final String cropName, final String programUUID,
+		final Integer studyId, final Integer variableTypeId) {
+		final BindingResult errors = new MapBindingResult(new HashMap<>(), StudyEntryServiceImpl.class.getName());
+
+		this.variableValidator.validateVariableTypeId(variableTypeId, errors);
+
+		if (!StringUtils.isEmpty(programUUID)) {
+			this.programValidator.validate(new ProgramDTO(cropName, programUUID), errors);
+		}
+
+		if (errors.hasErrors()) {
+			throw new ResourceNotFoundException(errors.getAllErrors().get(0));
+		}
+
+		this.studyValidator.validate(studyId, true);
+
+		return this.middlewareStudyEntryService.getStudyEntryDetails(cropName, programUUID, studyId, variableTypeId);
+	}
+
+	@Override
+	public void importUpdates(final Integer studyId, final StudyEntryDetailsImportRequest studyEntryDetailsImportRequest) {
+		if (!org.fest.util.Collections.isEmpty(studyEntryDetailsImportRequest.getNewVariables())) {
+			final Integer datasetId = this.datasetService.getDatasets(
+				studyId, Collections.singleton(DatasetTypeEnum.PLOT_DATA.getId())).get(0).getDatasetId();
+			this.datasetService.addDatasetVariables(studyId, datasetId, studyEntryDetailsImportRequest.getNewVariables());
+		}
+
+		this.studyValidator.validate(studyId, true);
+		final List<StudyEntryDetailsValueMap> entryDetailsValues = studyEntryDetailsImportRequest.getData();
+		final Map<String, List<StockPropertyData>> entriesMap = entryDetailsValues.stream()
+			.collect(Collectors.toMap(StudyEntryDetailsValueMap::getEntryNumber, StudyEntryDetailsValueMap::getData));
+		this.studyEntryValidator.validateStudyContainsEntryNumbers(studyId, entriesMap.keySet());
+
+		// retrieve study entries dto
+		final StudyEntrySearchDto.Filter filter = new StudyEntrySearchDto.Filter();
+		filter.setEntryNumbers(new ArrayList<>(entriesMap.keySet()));
+		final List<StudyEntryDto> studyEntries =
+			this.middlewareStudyEntryService.getStudyEntries(studyId, filter, new PageRequest(0, Integer.MAX_VALUE));
+
+		studyEntries.forEach(entry ->
+			this.processStudyEntry(studyId, entriesMap.get(String.valueOf(entry.getEntryNumber())), entry)
+		);
+	}
+
+	private void processStudyEntry(final Integer studyId, final List<StockPropertyData> entriesDataList,
+		final StudyEntryDto entry) {
+		if (!org.fest.util.Collections.isEmpty(entriesDataList)) {
+			entriesDataList.forEach(variable -> {
+				variable.setStockId(entry.getEntryId());
+				if (variable.hasValue() && !variable.getValue().isEmpty()) {
+					this.studyEntryObservationService.createOrUpdateObservation(studyId, variable);
+				} else {
+					final Optional<StockProperty> stockPropertyOptional = this.middlewareStudyEntryService.getByStockIdAndTypeId(
```

Comment:
To improve performance, I think it's better to retrieve the stock properties in one go rather than querying them one by one. I would convert the stock properties to MultiKeyMap (stockid and typeid as keys) and use that to check if a record with stockid+typeid combination exists.